src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java

```diff
@@ -249,6 +259,9 @@ public void popOutMessagesWithQosZero() {
     }
 
     private void removeMessagesWithQosZero(boolean needToCheckCurSpoolerSize) {
+        if (!qos0MessageCheckRequired.get()) {
+            return;
+        }
```

Comment:
side note: i think there's a concurrency issue. `removeMessagesWithQosZero` is called 1) when MqttClient connection is interrupted and 2) during spool addMessage.   The latter is synchronized, but the former is not.   Technically a different issue so can be addressed separately, just wanted to call it out

Reply:
Thanks...Nice catch, I can make a note and add it to next PR.

Reply:
still a problem?

Reply:
Umm..yes, we don't know if some add/remove message could change conditions midway. I'm not sure, will it hurt adding synchronized though?

Reply:
If `needToCheckCurSize` is false, maybe just don't check the flag. Always do the cleanup.

Reply:
We are looping through the whole database (even if we don't have QOS0 messages), if we don't check the flag in that case.

Reply:
Yes, but that's only called on disconnect, not on every single message. So probably not a big deal.

Reply:
To confirm we are concerned about the additional delay it will cause due to being synchronized right? In that case, wanted to mention...`addMessage` anyways is synchronized (the other place which calls `removeMessagesWithQosZero`).

Reply:
I'm just concerned about a potential deadlock, shouldn't happen, but adding synchronized where it wasn't before does have potential problems.

Reply:
Given that we are using an atomic boolean, this should still be correct

Reply:
If we exit early and don't remove 1 qos 0 message that doesn't seem too bad anyway.